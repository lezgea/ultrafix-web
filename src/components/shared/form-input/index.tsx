import React from 'react';
import { FieldErrors, UseFormRegister } from 'react-hook-form';


interface IFormInputProps {
    label: string;
    type: string;
    placeholder: string;
    errors: FieldErrors;
    register: UseFormRegister<any>;
    name: string;
    icon?: JSX.Element;
    labelStyle?: React.CSSProperties;
    inputStyle?: React.CSSProperties;
    onClickIcon?: () => void;
}

export const FormInput: React.FC<IFormInputProps> = (props) => {
    let {
        label,
        type,
        placeholder,
        errors,
        icon,
        register,
        name,
        labelStyle,
        inputStyle,
        onClickIcon,
    } = props;

    return (
        <div className="relative">
            <div className="relative">
                <label htmlFor={name} className="block text-gray-700 mb-2" style={labelStyle}>
                    {label}
                </label>
                <input
                    type={type}
                    id={name}
                    placeholder={placeholder}
                    className={`w-full h-[50px] bg-gray-50 px-5 py-2 pr-12 border ${errors[name] ? 'ring-2 ring-red-500' : 'border-gray-300'
                        } rounded-xl focus:outline-none focus:ring-2 focus:ring-primaryLight transition duration-200 ease-in-out transform`}
                    style={inputStyle}
                    {...register(name)}
                />
                <div className="absolute h-[50px] right-5 bottom-0 flex items-center" onClick={onClickIcon}>
                    {icon}
                </div>
            </div>
            {
                errors[name] && (
                    <p className="text-red-500 text-sm mt-1">{(errors[name]?.message as string) || ''}</p>
                )
            }
        </div>
    )
}
