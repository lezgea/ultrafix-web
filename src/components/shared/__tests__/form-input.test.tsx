import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import { FieldErrors, UseFormRegister } from 'react-hook-form';
import { FormInput } from '../form-input';

// Mock register function
const mockRegister = jest.fn((name: string) => ({
    name,
    onBlur: jest.fn(),
    onChange: jest.fn(),
    ref: jest.fn(),
}));

describe('FormInput Component', () => {
    it('renders without errors', () => {
        const { asFragment } = render(
            <FormInput
                label="Name"
                type="text"
                placeholder="Enter your name"
                errors={{}}
                register={mockRegister as unknown as UseFormRegister<any>}
                name="name"
            />
        );

        expect(screen.getByLabelText('Name')).toBeInTheDocument();
        expect(screen.getByPlaceholderText('Enter your name')).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    it('displays error message when there is an error', () => {
        const errors: FieldErrors = {
            name: { type: 'required', message: 'Name is required' },
        };

        const { asFragment } = render(
            <FormInput
                label="Name"
                type="text"
                placeholder="Enter your name"
                errors={errors}
                register={mockRegister as unknown as UseFormRegister<any>}
                name="name"
            />
        );

        expect(screen.getByText('Name is required')).toBeInTheDocument();
        expect(screen.getByLabelText('Name')).toHaveClass('ring-2 ring-red-500');
        expect(asFragment()).toMatchSnapshot();
    });

    it('renders with an icon and calls onClickIcon when clicked', () => {
        const mockOnClickIcon = jest.fn();
        const icon = <span>Icon</span>;

        const { asFragment } = render(
            <FormInput
                label="Password"
                type="password"
                placeholder="Enter your password"
                errors={{}}
                register={mockRegister as unknown as UseFormRegister<any>}
                name="password"
                icon={icon}
                onClickIcon={mockOnClickIcon}
            />
        );

        expect(screen.getByText('Icon')).toBeInTheDocument();
        fireEvent.click(screen.getByText('Icon'));
        expect(mockOnClickIcon).toHaveBeenCalledTimes(1);
        expect(asFragment()).toMatchSnapshot();
    });
});

