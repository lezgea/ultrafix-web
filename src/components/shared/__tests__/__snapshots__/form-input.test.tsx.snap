// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormInput Component displays error message when there is an error 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      class="relative"
    >
      <label
        class="block text-gray-700 mb-2"
        for="name"
      >
        Name
      </label>
      <input
        class="w-full h-[50px] bg-gray-50 px-5 py-2 pr-12 border ring-2 ring-red-500 rounded-xl focus:outline-none focus:ring-2 focus:ring-primaryLight transition duration-200 ease-in-out transform"
        id="name"
        name="name"
        placeholder="Enter your name"
        type="text"
      />
      <div
        class="absolute h-[50px] right-5 bottom-0 flex items-center"
      />
    </div>
    <p
      class="text-red-500 text-sm mt-1"
    >
      Name is required
    </p>
  </div>
</DocumentFragment>
`;

exports[`FormInput Component renders with an icon and calls onClickIcon when clicked 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      class="relative"
    >
      <label
        class="block text-gray-700 mb-2"
        for="password"
      >
        Password
      </label>
      <input
        class="w-full h-[50px] bg-gray-50 px-5 py-2 pr-12 border border-gray-300 rounded-xl focus:outline-none focus:ring-2 focus:ring-primaryLight transition duration-200 ease-in-out transform"
        id="password"
        name="password"
        placeholder="Enter your password"
        type="password"
      />
      <div
        class="absolute h-[50px] right-5 bottom-0 flex items-center"
      >
        <span>
          Icon
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`FormInput Component renders without errors 1`] = `
<DocumentFragment>
  <div
    class="relative"
  >
    <div
      class="relative"
    >
      <label
        class="block text-gray-700 mb-2"
        for="name"
      >
        Name
      </label>
      <input
        class="w-full h-[50px] bg-gray-50 px-5 py-2 pr-12 border border-gray-300 rounded-xl focus:outline-none focus:ring-2 focus:ring-primaryLight transition duration-200 ease-in-out transform"
        id="name"
        name="name"
        placeholder="Enter your name"
        type="text"
      />
      <div
        class="absolute h-[50px] right-5 bottom-0 flex items-center"
      />
    </div>
  </div>
</DocumentFragment>
`;
